infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,613 @@
+        this.timeoutResponse = timeoutResponse;
+    }
+
+    /**
+     * Interfaces between the LogReplicationServer (msg handler), and the transport layer
+     * @param response Log replication response message
+     */
+    @Override
+    public void sendResponse(CorfuMessage.ResponseMsg response) {
+        log.trace("Ready to send response {}", response.getPayload().getPayloadCase());
+        LogReplicationSession session = response.getHeader().getSession();
+        if (isConnectionInitiator(session)) {
+            if(sessionManager.getIncomingSessions().contains(session)) {
+                if (sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().isPresent()) {
+                    clientChannelAdapter.send(sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().get(), response);
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(
+                            new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_LOSS));
+                }
+            }
+        } else {
+            try {
+                this.serverChannelAdapter.send(response);
+                log.trace("Sent response: {}", response);
+            } catch (IllegalArgumentException e) {
+                log.warn("Illegal response type. Ignoring message.", e);
+            }
+        }
+    }
+
+    /**
+     * Receive Corfu Message from the Channel Adapter for further processing
+     *
+     * @param msg received corfu message
+     */
+    @Override
+    public void receive(CorfuMessage.ResponseMsg msg) {
+        try {
+            LogReplicationSession session = msg.getHeader().getSession();
+
+            // If it is a Leadership Loss Message re-trigger leadership discovery
+            if (msg.getPayload().getPayloadCase() == CorfuMessage.ResponsePayloadMsg.PayloadCase.LR_LEADERSHIP_LOSS) {
+                String nodeId = msg.getPayload().getLrLeadershipLoss().getNodeId();
+                if (isOutgoingSession(session)) {
+                    this.sessionToRuntimeFSM.get(session)
+                            .input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS, nodeId));
+                } else {
+                    sessionToSinkVerifyLeadership.get(session)
+                            .input(new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_LOSS, nodeId));
+                }
+                return;
+            }
+
+            if (msg.getPayload().getPayloadCase().equals(CorfuMessage.ResponsePayloadMsg.PayloadCase.LR_SUBSCRIBE_MSG)) {
+                String RemoteLeaderId = msg.getPayload().getLrSubscribeMsg().getSinkLeaderNodeId();
+                // Start runtimeFSM
+                sessionToRuntimeFSM.get(session).setRemoteLeaderNodeId(RemoteLeaderId);
+                sessionToRuntimeFSM.get(session).start();
```

Comment:
what will happen if the runtime fsm was started, then the connection with the sink cluster went down and came back up after some time.  The start() will get executed again and another worker will start polling the event queue, right?

Reply:
Correct for most, except for the last part...the same worker thread will continue polling the event queue. When connection is lost, onConnectionDown() gets triggered, and the FSM goes to verifyRemoteLeader state.
There depending on if it is the connection initiator or not, it either sends the leadership query or waits until the `remoteLeaderNodeId` is set